src/xcode/ENA/ENA/Source/Client/Security/CoronaWarnURLSessionDelegate.swift

```diff
@@ -0,0 +1,143 @@
+// file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+//
+
+import Foundation
+import CommonCrypto
+import CryptoKit
+
+final class CoronaWarnURLSessionDelegate: NSObject {
+	// MARK: Known Public Key Storage
+	/// A dictionary containing a mapping of the host to the SHA256 public key string
+	private let domainPublicKeyHashes: [String: String]
+	/// Whitelist of domains we do not pin the public key for. Currently just the distribution URL
+	private let whitelist: [String]
+
+	// MARK: Creating a Delegate
+	override init() {
+		guard
+			let publicKeyDict = Bundle.main.readPlistDict(name: "PublicKeys"),
+			let hostWhitelist = Bundle.main.readPlistAsArr(name: "HostWhitelist")
+		else {
+			preconditionFailure("Could not load PublicKeys.plist for public key pinning!")
+		}
+
+		domainPublicKeyHashes = publicKeyDict
+		whitelist = hostWhitelist
+	}
+}
+
+extension CoronaWarnURLSessionDelegate: URLSessionDelegate {
+	func urlSession(
+		_ session: URLSession,
+		didReceive challenge: URLAuthenticationChallenge,
+		completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
+	) {
+		func reject() { completionHandler(.cancelAuthenticationChallenge, /* credential */ nil) }
+
+		guard !checkWhitelist(for: challenge.protectionSpace.host) else {
+			completionHandler(.performDefaultHandling, nil)
+			return
+		}
+
+		// `serverTrust` not nil implies that authenticationMethod == NSURLAuthenticationMethodServerTrust
+		guard
+			let trust = challenge.protectionSpace.serverTrust,
+			let localPublicKey = key(for: challenge.protectionSpace.host),
+			!localPublicKey.isEmpty
+		else {
+			// Reject all requests that we do not have a public key to pin for
+			reject()
+			return
+		}
+
+		// We discard the returned status code (OSStatus) because this is also how
+		// Apple is doing it in their official sample code – see [0] for more info.
+		SecTrustEvaluateAsyncWithError(trust, .main) { trust, isValid, error in
+			func accept() { completionHandler(.useCredential, URLCredential(trust: trust)) }
+
+			guard isValid else {
+				logError(message: "Server certificate is not valid. Rejecting challenge!")
+				reject()
+				return
+			}
+
+			guard error == nil else {
+				logError(message: "Encountered error when evaluating server trust challenge, rejecting!")
+				reject()
+				return
+			}
+
+			// Our landscape has a certificate chain with three certificates.
+			// We want to get the intermediate certificate, in our case the second.
+			guard
+				SecTrustGetCertificateCount(trust) >= 1,
```

Comment:
Would this be >= 2 because we have the server certificate as well as the intermediate certificate --> 2 certificates